### PR TITLE
Fix/31856: User is signed out

### PR DESCRIPTION
**Pull request by DylanDylann**
### Fixed Issues

$ https://github.com/Expensify/App/issues/31856
PROPOSAL: https://github.com/Expensify/App/issues/31856#issuecomment-1826217940

### Tests

1. Open OD (has not signed in yet)
2. Click on 'Email' > enter email (that has not registered yet) in input field > click on 'Next' button > click on 'Join'
3. It will navigate to ND
4. Navigate to Profile settings > click on 'Contact method' > click on your email
5. Click on 'Didn't receive a magic code?' link > copy magic code from email > paste the magic code in the input field
6. Verify that user will be signed out right after that.
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Open OD (has not signed in yet)
2. Click on 'Email' > enter email (that has not registered yet) in input field > click on 'Next' button > click on 'Join'
3. It will navigate to ND
4. Navigate to Profile settings > click on 'Contact method' > click on your email
5. Click on 'Didn't receive a magic code?' link > copy magic code from email > paste the magic code in the input field
6. Verify that user will be signed out right after that.
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js [hunk 1]
```diff
@@ -119,12 +120,21 @@ class ContactMethodDetailsPage extends Component {
...
+        const validatedDate = lodashGet(this.props.loginList, [contactMethod, 'validatedDate']);
+        const prevValidatedDate = lodashGet(prevProps.loginList, [contactMethod, 'validatedDate']);
 
+        const loginData = lodashGet(this.props.loginList, contactMethod, {});
+        const isDefaultContactMethod = this.props.session.email === loginData.partnerUserID;
+        const isFailedAddContactMethod = Boolean(lodashGet(loginData, 'errorFields.addedLogin'));
+        const shouldSignOut = isDefaultContactMethod && !loginData.validatedDate && !isFailedAddContactMethod;
         // Navigate to methods page on successful magic code verification
         // validatedDate property is responsible to decide the status of the magic code verification
         if (!prevValidatedDate && validatedDate) {
```
[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:133]: No. Once we call ValidateSecondaryLogin API, the validatedDate is defined, and the prevValidatedDate is undefined

[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:133]: Oh. Sorry @Ollyws, my bad. Just updated the shouldSignout logic.

[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:133]: > You can test the solution by setting the loginList onyx key with an empty validatedDate to emulate an account that was redirected from OldDot.

I will do it.

> we will never get past the if (!prevValidatedDate && validatedDate)

You `ve tried it based on this [suggestion](https://github.com/Expensify/App/issues/31856#issuecomment-1847827786), right?

[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:133]: > The idea is that we are checking for an account with no validatedDate in order to determine that it was redirected by OldDot and never validated, correct?

The account that is redirected from OD and never validated need to match the condition: it is the current ```session.email``` and the ```validatedDate``` is undefined

[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:133]: >  it looks like validatedDate is empty

Yes. After we fill in the magic code, the API ValidateSecondaryLogin is called, so the validatedDate is not empty

[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:133]: You can see the screenshot below:

[Screencast from 13-12-2023 23:52:56.webm](https://github.com/Expensify/App/assets/141406735/812c428f-2fe8-4745-950b-afd8bdd58985)

[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:133]: @Ollyws Also pls note that, we only sign out in this case: OD => ND => click on profile => click on contact method => click on the default contact method =>  enter magic code. In other cases, user is not signed out.

[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:133]: - I found out that in the case that leads to this bug (This bug that I mentioned is: User is signed out), it need to match all the bellow condition:
1. The selected contact method is the session.email
2. The selected contact method has validatedDate field is undefined before calling the API ValidateSecondaryLogin
3. The selected contact method has validatedDate field is defined after the API ValidateSecondaryLogin is called successfully

- So, the ```shouldSignOut``` is used to match the above conditions

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js [hunk 2]
```diff
@@ -119,12 +120,21 @@ class ContactMethodDetailsPage extends Component {
...
+        const prevValidatedDate = lodashGet(prevProps.loginList, [contactMethod, 'validatedDate']);
 
+        const loginData = lodashGet(this.props.loginList, contactMethod, {});
+        const isDefaultContactMethod = this.props.session.email === loginData.partnerUserID;
+        const isFailedAddContactMethod = Boolean(lodashGet(loginData, 'errorFields.addedLogin'));
+        const shouldSignOut = isDefaultContactMethod && !isFailedAddContactMethod;
         // Navigate to methods page on successful magic code verification
         // validatedDate property is responsible to decide the status of the magic code verification
         if (!prevValidatedDate && validatedDate) {
+            if (shouldSignOut) {
```
[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:135]: Updated

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js [hunk 3]
```diff
@@ -119,12 +120,22 @@ class ContactMethodDetailsPage extends Component {
...
+        const validatedDate = lodashGet(this.props.loginList, [contactMethod, 'validatedDate']);
+        const prevValidatedDate = lodashGet(prevProps.loginList, [contactMethod, 'validatedDate']);
 
+        const loginData = lodashGet(this.props.loginList, contactMethod, {});
+        const isDefaultContactMethod = this.props.session.email === loginData.partnerUserID;
+        const shouldSignOut = isDefaultContactMethod;
         // Navigate to methods page on successful magic code verification
         // validatedDate property is responsible to decide the status of the magic code verification
         if (!prevValidatedDate && validatedDate) {
+            // If the selected contactMethod is the current session.login and the account is unvalidated, the current authToken is invalid after the successful magic code verification
```
[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js]: Sure. Updated it

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js [hunk 4]
```diff
@@ -119,12 +120,23 @@ class ContactMethodDetailsPage extends Component {
...
     componentDidUpdate(prevProps) {
-        const validatedDate = lodashGet(this.props.loginList, [this.getContactMethod(), 'validatedDate']);
-        const prevValidatedDate = lodashGet(prevProps.loginList, [this.getContactMethod(), 'validatedDate']);
+        const contactMethod = this.getContactMethod();
+        const validatedDate = lodashGet(this.props.loginList, [contactMethod, 'validatedDate']);
+        const prevValidatedDate = lodashGet(prevProps.loginList, [contactMethod, 'validatedDate']);
 
+        const loginData = lodashGet(this.props.loginList, contactMethod, {});
+        const isDefaultContactMethod = this.props.session.email === loginData.partnerUserID;
+        const shouldSignOut = isDefaultContactMethod;
```
[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js]: Updated

**Review by Ollyws — APPROVED** [review]
As I don't have access to a local version of oldDot, I tested this by initially clearing `validatedDate` in Onyx, then updating `validatedDate` to a new date in Onyx when submitting the validation code to simulate the server response.

**Comment by thienlnam** [comment]
Thanks - I'll test this on my end now

**Comment by thienlnam** [comment]
https://github.com/Expensify/App/assets/30609178/fd957f9c-7f2f-47e4-8921-ef24037eb697

Nice, looks like it works as intended!

**Review by thienlnam — APPROVED** [review]

**Pull request merged**